# Module 4 Practice - Python Fundamentals
## Working with Files

<font size="5" color="#00A0B2"  face="verdana"> <B>Student will be able to</B></font>  
 
- Import files into Jupyter notebooks  
- Open and **`.read()`** local files in memory   
- Use **`.read()`** to read a specific number of characters  
- Use **`.readlines()`** to read data from file as a **list** of lines  
- Use **`.readlines()`** to read data from file as a **list** of lines   
- Use **`.readline()`** to read data from file a line at a time   
- Use **`.strip()`** to remove new line characters  
- Use `.write()` to write data to a new local file  
- Use **`.seek()`** to set file read or write location  
- Use file append mode  


# &nbsp;
<font size="6" color="#B24C00"  face="verdana"> <B>Task 1</B></font>

## Order the Rainbow
#### Open the rainbow file then put in a list and print in alphabetical order
Download and open the file. 
- Download list of rainbow colors, as `rainbow.txt`, using `curl` 
  https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/rainbow   
- Open rainbow.txt in read mode using a variable: rainbow_file   
 

In [1]:
# [ ] import https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/rainbow as rainbow.txt

!curl https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/rainbow -o rainbow.txt

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100    43  100    43    0     0    228      0 --:--:-- --:--:-- --:--:--   232



- Read rainbow_file as a list variable: rainbow_colors using `.readlines()`    


In [3]:
# [ ]  Open rainbow.txt in read mode & read as list with .readlines()

rainbow_txt = open('rainbow.txt', 'r')
rain = rainbow_txt.readlines()


 
1. Sort the rainbow_lines list alphabetically.  
2. Print each line of rainbow_lines by iterating the sorted list.  
3. Close rainbow_file.  

In [5]:
# [ ] sort rainbow_colors list, iterate the list to print each color

rain.sort()
print(rain)
rainbow_txt.close()

['blue\n', 'green\n', 'indigo\n', 'orange\n', 'red\n', 'violet\n', 'yellow\n']


# &nbsp;
<font size="6" color="#B24C00"  face="verdana"> <B>Task 2</B></font>

## The Weather
Create a program that reads from a file to display city name and average temperature in Celsius.   

&nbsp;  
- use `!curl` to download https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/world_temp_mean.csv as `mean_temp.txt`  
  


In [6]:
# [ ] The Weather: import world_mean_team.csv as mean_temp.txt
!curl  https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/world_temp_mean.csv -o mean_temp.txt


  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   222  100   222    0     0   1263      0 --:--:-- --:--:-- --:--:--  1290


1. Open the file in `'r'` mode.  
2. Read the first line of text into a variable called: `  headings` and `print()`.  
3. Convert `headings` to a list using **`.split(',')`** which splits on each comma, `print()` the list.


In [14]:
# [ ] The Weather: open file, read/print first line, convert line to list (splitting on comma)

file = open('mean_temp.txt','r')
headings = file.read()
print(headings)
hdsplit= headings.split(',')
print(hdsplit)


city,country,month ave: highest high,month ave: lowest low
Beijing,China,30.9,-8.4
Cairo,Egypt,34.7,1.2
London,UK,23.5,2.1
Nairobi,Kenya,26.3,10.5
New York City,USA,28.9,-2.8
Sydney,Australia,26.5,8.7
Tokyo,Japan,30.8,0.9

['city', 'country', 'month ave: highest high', 'month ave: lowest low\nBeijing', 'China', '30.9', '-8.4\nCairo', 'Egypt', '34.7', '1.2\nLondon', 'UK', '23.5', '2.1\nNairobi', 'Kenya', '26.3', '10.5\nNew York City', 'USA', '28.9', '-2.8\nSydney', 'Australia', '26.5', '8.7\nTokyo', 'Japan', '30.8', '0.9\n']


#### use a while loop to read the remaining lines from the file  
  1. Assign remaining lines to a **`city_temp`** variable.  
  2. Convert the city_temp to a list using **`.split(',')`** for each **`.readline()`** in the loop.  
  3. Print each city & the highest monthly average temperature.  
  4. Close mean_temps.  

>Tips & Hints:   
- Use the print output of **`headings`** to determine the city_temp indexes to use.  
- "month ave: highest high" for Beijing is 30.9 Celsius.  
- Convert `city_temp` to lists with `.split(',')`.  

In [16]:
# [ ] The Weather: use while loop to print city and highest monthly average temp in celsius

city_temp = file.readline()
cityinfo = city_temp.split(",")
while city_temp:
    print(cityinfo[0]+': '+cityinfo[2])
    city_temp = file.readline()
    cityinfo = city_temp.split(',')
file.close()

# &nbsp;  
<font size="6" color="#B24C00"  face="verdana"> <B>Task 3</B></font>  

## Random pi  guessing
Create random appearing numbers by reading digits of pi 
*Note: only "appears" random*

- Download https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/digits_of_pi  as pi.txt

In [17]:
# [ ] use curl to download https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/digits_of_pi as pi.txt

!curl https://raw.githubusercontent.com/MicrosoftLearning/intropython/master/digits_of_pi -o pi.txt

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   303  100   303    0     0   2197      0 --:--:-- --:--:-- --:--:--  2227


#### Set up the project files and intitial values  
1. Open pi.txt in read mode, the file has a single line of text "3.14....".  
2. Get user name as input and say "hi".  
3. Use the length of `name` for variable called `seed`.  
4. Use `.seek()` with the value of `seed` to set the initial pointer location reading the file.  
5. Create a variable `digit` and assign it the value of reading one character from the file.  
6. Get `guess` variable value from users `input`  - "enter a single digit guess or "q" to quit".  
7. Initialize `correct` and `wrong` counter variables to **`0`**  (zero).  

In [26]:
# [ ] Set up the project files and initial values
pi = open('pi.txt','r')
user= input("What's your name?")
print('Hi', user)
seed = len(user)
pi.seek(seed)
digit = pi.read(1)
guess = input('Enter a single digit or \'q\' to quit: ')
correct = 0
wrong = 0

Hi keoni


#### Create a while loop that tests that `guess` is a *digit* string
then in the loop:
1. if `digit` ( read from pi file) is "**.**" read the next character for digit  
2. else if `digit` is "\n" increment `seed` and use `seed` to set the pointer uing .`seek()`  
3. else see if `guess` is equal to `digit`   
  a. if `guess` equals `digit`: print "correct" and increment the varible named `correct`  
  b. if `guess` not equal `digit`: print "incorrect" and increment the variable named `wrong`  
  
**end the while loop** when user enters any non-digit(s) for `guess`, like "q".  
- Print `correct` and `wrong` values within a message to the user.  
- Close the pi file.  

In [28]:


while True:
    guess = input("Enter a single digit or \"q\" to quit")

    if digit == '.':
        digit =pi.read(1)
    elif digit == '\n':
        seed = int(len(user))+1
        pi.seek(seed)
    elif guess == digit:
        print('correct')
        correct += 1
    elif guess == 'q':
        break
    else:
        print('incorrect')
        wrong +=1


print(user +"'s " + "score: ")
print('correct:', correct)
print('incorrect:', wrong)

incorrect
incorrect
incorrect
correct
incorrect
incorrect
incorrect
incorrect
incorrect
incorrect
correct
incorrect
correct
correct
correct
keoni's score: 
correct: 5
incorrect: 1401093


[Terms of use](http://go.microsoft.com/fwlink/?LinkID=206977) &nbsp; [Privacy & cookies](https://go.microsoft.com/fwlink/?LinkId=521839) &nbsp; © 2017 Microsoft